packages/patrol_cli/lib/src/features/test/ios_test_backend.dart

```diff
@@ -47,110 +50,182 @@ class IOSAppOptions {
-      task.fail('Failed to run `flutter build` for $targetName');
-      throw Exception('flutter build exited with code $exitCode');
-    }
+  @override
+  Future<void> build(IOSAppOptions options, Device device) async {
+    await _disposeScope.run((scope) async {
+      final subject = '${options.desc} for ${device.platformDescription}';
+      final task = _logger.task('Building $subject');
 
-    final xcodebuildProcess = await _processManager.start(
-      options.toXcodebuildInvocation(device),
-      runInShell: true,
-      workingDirectory: _fs.currentDirectory.childDirectory('ios').path,
-    );
-
-    xcodebuildProcess.stdout.listen((rawMsg) {
-      final msg = systemEncoding.decode(rawMsg).trim();
-      _logger.detail('\t$msg');
-    }).disposedBy(_disposeScope);
-
-    xcodebuildProcess.stderr.listen((rawMsg) {
-      final msg = systemEncoding.decode(rawMsg).trim();
-      _logger.err('\t$msg');
-    }).disposedBy(_disposeScope);
-
-    exitCode = await xcodebuildProcess.exitCode;
-    if (exitCode == 0) {
-      task.complete('Built and ran app for $targetName on ${device.name}');
-    } else {
-      task.fail('Failed to build/run app for $targetName on ${device.name}');
-      throw Exception('xcodebuild exited with code $exitCode');
-    }
+      Process process;
+
+      // flutter build ios --config-only
+
+      var flutterBuildKilled = false;
+      process = await _processManager.start(
+        options.toFlutterBuildInvocation(device),
+        runInShell: true,
+      );
+      scope.addDispose(() async {
+        process.kill();
+        flutterBuildKilled = true; // `flutter build` has exit code 0 on SIGINT
+      });
+      process.listenStdOut((l) => _logger.detail('\t$l')).disposedBy(scope);
+      process.listenStdErr((l) => _logger.err('\t$l')).disposedBy(scope);
+      var exitCode = await process.exitCode;
+      if (exitCode != 0) {
+        final cause = '`flutter build ios` exited with code $exitCode';
+        task.fail('Failed to build $subject ($cause)');
+        throw Exception(cause);
+      } else if (flutterBuildKilled) {
+        const cause = '`flutter build ios` was interrupted';
+        task.fail('Failed to build $subject ($cause)');
+        throw Exception(cause);
+      }
+
+      // xcodebuild build-for-testing
+
+      process = await _processManager.start(
+        options.buildForTestingInvocation(device),
+        runInShell: true,
+        workingDirectory: _fs.currentDirectory.childDirectory('ios').path,
+      )
+        ..disposedBy(scope);
+      process.listenStdOut((l) => _logger.detail('\t$l')).disposedBy(scope);
+      process.listenStdErr((l) => _logger.err('\t$l')).disposedBy(scope);
+      exitCode = await process.exitCode;
+      if (exitCode == 0) {
+        task.complete('Completed building $subject');
+      } else {
+        final cause = 'xcodebuild exited with code $exitCode';
+        task.fail('Failed to build $subject ($cause)');
+        throw Exception(cause);
```

Comment:
Have you thought about creating a child-class for exceptions indicating failures in app/app backend processes?

Reply:
I'm constantly putting it for later, haha

I'm going to hesitantly say that for now, it's okay - the type of exception that is thrown doesn't matter at all to the code that catches the exception.

What is important is that _anything_ was thrown - then the test run is marked as failed.